### ISSUE TITLE
phantomjs kill failing TypeError

### ISSUE THREAD

**Issue by sashacm** [prerender/prerender#246]
I just upgraded our setup to 3.22.1 and started seeing what looked like a memory leak. I restarted things a couple times to recover, but kept watching memory grow steadily. Digging in this morning, I see that there are hundreds of PhantomJS processes in the background, and that the logs indicate failures when trying to kill them:

```
/usr/lib/node_modules/prerender/lib/server.js:493
    require('tree-kill')(this.phantom.process.pid, 'SIGTERM');
                                     ^
TypeError: Cannot read property 'process' of null
    at Object.server._killPhantomJS (/usr/lib/node_modules/prerender/lib/server.js:493:38)
    at Object.server._sendResponse (/usr/lib/node_modules/prerender/lib/server.js:487:16)
    at /usr/lib/node_modules/prerender/lib/server.js:447:37
    at Gzip.onEnd (zlib.js:167:5)
    at Gzip.emit (events.js:117:20)
    at _stream_readable.js:944:16
    at process._tickCallback (node.js:448:13)
```
I see that this kill line was changed recently, and i switched back to the old approach:
```
    //require('tree-kill')(this.phantom.process.pid, 'SIGTERM');
      require('tree-kill')(this.options.worker.process.pid, 'SIGTERM');
```

It's only been a little while, but things seem happier. It may be something in our setup and I haven't had a chance to dig in deeper, but figured I'd log this in case it's a broader issue with the new kill approach.

We're running with the prerender redis cache plugin, but otherwise a pretty vanila/OOTB setup. Default number of workers (2 core boxes), default iterations.

**Comment by thoop** [comment]
Hmm, we run with the new code and it works better than before in keeping the number of spare phantomjs processes away.

It looks like in your example `this.phantom` is null. Did you change anything else in the code? Can you verify that `this.phantom` gets set to the new phantomjs instance when it's created?

**Comment by sashacm** [comment]
Sorry, haven't had a chance to follow up. I checked out your repo (master, since it doesn't look like anything's been added since you versioned 3.22.1) and diffed against what we have running (npm installed and very lightly puppet-modified). The only differences:
- 2 urls added into lib/resources/blocked-resources.json
- modified server.js to append "server.use(require('prerender-redis-cache'));"

Not sure if the redis cache might make a difference, but it's the only obvious/meaningful difference in play.

I'll try to do a bit more debugging when I get a chance to see what's happening that might cause ```this.phantom``` to not be set.

**Comment by thoop** [comment]
Great, thanks!

**Comment by maxlath** [comment]
I'm having the same issue: phantomjs throws the same error as @sashacm and end up eating all the memory.
Prerender 3.22.1 (with some [custom plugins](https://github.com/inventaire/prerender))
Ubuntu 15.10

**Comment by S-YOU** [comment]
which version of nodejs you using btw? @sashacm

**Comment by maxlath** [comment]
I'm still on v0.10.40 as phantomjs wasn't ready for 4.2 last time I tried

**Comment by S-YOU** [comment]
I see, I had similar experience with node5.0.0 and weak module, getting hundreds of PhantomJS processes, and server eventually freezed, but looks like seems ok on node 5.1.0.

**Comment by sashacm** [comment]
We're on v0.10.37. Haven't had a chance to circle back to debug further, unfortunately. Always seems to go that way once you find a (temporary?) solution that makes the urgent problem go away.

**Comment by thoop** [comment]
@maxlath Are you seeing `this.phantom` being null? Did you change anything else in the code? Can you verify that this.phantom gets set to the new phantomjs instance when it's created?

We don't see this on our instances.

**Comment by maxlath** [comment]
@thoop `this.phantom` is most of the times not `null` (I checked it by doing a `console.log(this.phantom)` in `server._killPhantomJS`), but sometimes somehow it crashes with `TypeError: Cannot read property 'process' of null`. Could it be that in some circumstances, the `onExit` is called before `killPhantomJS` has it seem to be the only piece of code doing a `_this.phantom = null`?

I haven't modified anything in [my lib/server.js](https://github.com/inventaire/prerender/blob/master/lib/server.js) file

**Comment by thoop** [comment]
I guess it's possible but it's weird. If phantom is null or the new phantom instance is different, then we abort the request (the case of onExit being called mid request). https://github.com/prerender/prerender/blob/master/lib/server.js#L300

But in @sashacm's case they weren't being cleaned up. Are you seeing the phantom instances get cleaned up or are they hanging around?

**Comment by maxlath** [comment]
hanging around too, unless all the memory is eaten

**Comment by davidlgj** [comment]
We're having the same problem, sometimes `this.phantom` is just null. Seems to be a little more than 1/4 that can't be killed.

We're using the phantomjs2 branch with a couple of custom plugins and node 0.10.29, so maybe we mucked something up. But my guess its the old node version. Currently we modified server.js to do this instead:

```js
if (this.phantom && this.phantom.process && this.phantom.process.pid) {
    require('tree-kill')(this.phantom.process.pid, 'SIGTERM');
} else {
    require('tree-kill')(this.options.worker.process.pid, 'SIGTERM');
}
```

Seems stable :)
